File: clojure/raindrops/src/raindrops.clj
(ns raindrops)

(defn prime-factors [num]
  (loop [num num divisors [] current-div 2]
    (cond
      (> current-div num) divisors
      (= (/ num current-div) 1) (conj divisors current-div)
      (integer? (/ num current-div)) (recur (/ num current-div)
                                            (conj divisors current-div)
                                            current-div)
      :else (recur num divisors (inc current-div)))))

(def sounds {3 "Pling" 5 "Plang" 7 "Plong"})

(defn or-n [n string] (if (empty? string) (str n) string))

(defn convert [n]
  (->> n
       prime-factors
       set
       sort
       (map sounds)
       (apply str)
       (or-n n)))

File: ruby/raindrops/raindrops.rb
module Raindrops
  extend self

  def div_by(number, divisor)
    number % divisor == 0
  end

  def divisors
    {3 => "Pling", 5 => "Plang", 7 => "Plong"}
  end

  def convert(number)
    sounds = divisors.select do |div, sound|
      div_by(number, div)
    end.map do |div, sound|
      sound
    end.join

    if sounds.empty?
      number.to_s
    else
      sounds
    end
  end
end
